# Task 4.9 Intro to Data Visualizations With Python 

# Contents 

1. Importing Libraries 
2. Importing Customer data frame 
3. Data Wrangling 
4. Data Consistency Checks 
5. Merging Data frames 

# Task 4.6 Part 1 Questions 

1. Download the customer data set and add it to your “Original Data” folder.
2. Create a new notebook in your “Scripts” folder for part 1 of this task.
3. Import your analysis libraries, as well as your new customer data set as a dataframe.

# Importing Libraries 

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# Importing customers.csv data frame 

In [2]:
path = r'C:\Users\emily\OneDrive\Documents\Career Foundry 2022\Data Immersion\Achievment 4\Instacart Basket Analysis'

In [3]:
cust = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'customers.csv'))

# Data Wrangling 

4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [4]:
cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
cust.shape


(206209, 10)

#Renaming surnam column to be consistent with first name column

In [6]:
cust.rename(columns = {'Surnam':'Last Name'}, inplace = True)

#Renaming STATE column to State

In [7]:
cust.rename(columns = {'STATE':'State'}, inplace = True)

#Renaming n_dependants column to be more explanatory and correcting spelling error

In [8]:
cust.rename(columns = {'n_dependants':'no_of_dependents'}, inplace = True)

#Renaming Gender column for consistency

In [9]:
cust.rename(columns = {'Gender':'gender'}, inplace = True)

#Renaming State column for consistency

In [10]:
cust.rename(columns = {'State':'state'}, inplace = True)

#Renaming Age column for consistency

In [11]:
cust.rename(columns = {'Age':'age'}, inplace = True)

#Renaming fam_status column for clearer understanding

In [12]:
cust.rename(columns = {'fam_status':'family_status'}, inplace = True)

In [13]:
cust.head()

user_id First Name Last Name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   no_of_dependents family_status  income  
0                 3       married  165665  
1                 0        single   59285  
2                 2       married   99568  
3                 0        single   42049  
4                 1       married   40374

# Data Consistency Checks 

5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [14]:
cust.describe()

user_id            age  no_of_dependents         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

#Checking Data Types 

In [15]:
cust.dtypes

user_id              int64
First Name          object
Last Name           object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependents     int64
family_status       object
income               int64
dtype: object

#Changing 'user_id' from integer to string type 

In [16]:
cust['user_id'] = cust['user_id'].astype('str')

#Checking for missing values (NaN)

In [17]:
cust.isnull().sum()

user_id                 0
First Name          11259
Last Name               0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependents        0
family_status           0
income                  0
dtype: int64

In [18]:
cust_nan=cust[cust['First Name'].isnull()==True]

In [19]:
cust_nan

user_id First Name Last Name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        no_of_dependents family_status  income  
53                     2       married   41709  
73                     0        single   82518  
82                     3       married  117099  
99                     1       married  155673  
105                    3       married  151819  
...                  ...           ...     ...  
206038                 3       married   87783  
206044                 2       married  108488  
206090                 1       married   45275  
206162                 0        single   56325  
206171                 2       married   59222  

[11259 rows x 10 columns]

It is best to keep the 11,259 NaN values in the First_name columns as is since each customer has a user_id value assigned to them. This means we can still link the data frames together. 

#Checking for Mixed- data types 

In [20]:
for col in cust.columns.tolist():
  weird = (cust[[col]].applymap(type) != cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (cust[weird]) > 0:
    print (col)

First Name


First Name column contains mixed-data types. 

In [21]:
#Converting First Name column from Object to String 
cust['First Name'] = cust['First Name'].astype('str')

#Checking for Duplicates in Data frame 

In [22]:
cust[cust.duplicated()]

Empty DataFrame
Columns: [user_id, First Name, Last Name, gender, state, age, date_joined, no_of_dependents, family_status, income]
Index: []

No duplicates have been found in the customer.csv data frame. 

#Reducing data types to save RAM before merging the data frames together

In [23]:
cust['age'] = cust['age'].astype('int8')
cust['no_of_dependents'] = cust['no_of_dependents'].astype('int8')
cust['income'] = cust['income'].astype('int32')

# Merging Data frames 

6. Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the same data type! Hint: Make sure the key columns are the same data type! 

#Importing orders_products_merged_complete.pkl

In [24]:
ords_prods = pd.read_pickle(os.path.join(path, 'Data','Prepared Data','orders_products_merged_complete.pkl'))

In [25]:
ords_prods.head()


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order_placed  product_id  add_to_cart_order  reordered  \
0                           NaN         196                  1          0   
1                          15.0         196                  1          1   
2                          21.0         196                  1          1   
3                          29.0         196                  1          1   
4                          28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day        busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy      Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy  Least busiest days   
2         Soda  ...  Mid-range product  Regularly busy  Least busiest days   
3         Soda  ...  Mid-range product      Least busy  Least busiest days   
4         Soda  ...  Mid-range product      Least busy  Least busiest days   

      Most_orders max_order  loyalty_flag avg_price  spending_flag  \
0  Average orders        10  New customer  6.367797    Low spender   
1  Average orders        10  New customer  6.367797    Low spender   
2     Most orders        10  New customer  6.367797    Low spender   
3  Average orders        10  New customer  6.367797    Low spender   
4     Most orders        10  New customer  6.367797    Low spender   

  median_order_frequency         frequency_flag  
0                   20.5  Non-frequent customer  
1                   20.5  Non-frequent customer  
2                   20.5  Non-frequent customer  
3                   20.5  Non-frequent customer  
4                   20.5  Non-frequent customer  

[5 rows x 23 columns]

#Checking the size of current dataframe 

In [26]:
ords_prods.shape

(32404859, 23)

#Checking data type for key column = user_id

In [27]:
ords_prods.dtypes

order_id                          int64
user_id                           int64
order_number                      int64
orders_day_of_week                int64
order_hour_of_day                 int32
days_since_last_order_placed    float64
product_id                        int64
add_to_cart_order                 int64
reordered                         int64
product_name                     object
aisle_id                          int64
department_id                     int64
prices                          float64
price_range_loc                  object
busiest_day                      object
busiest_days                     object
Most_orders                      object
max_order                         int64
loyalty_flag                     object
avg_price                       float64
spending_flag                    object
median_order_frequency          float64
frequency_flag                   object
dtype: object

#Changing user_id column from integer to string

In [28]:
ords_prods['user_id'] = ords_prods['user_id'].astype('str')

Reducing space for more RAM

In [29]:
ords_prods['product_id'] = ords_prods['product_id'].astype('int16')

In [30]:
ords_prods['order_number'] = ords_prods['order_number'].astype('int8')

In [31]:
ords_prods['orders_day_of_week'] = ords_prods['orders_day_of_week'].astype('int8')

In [32]:
ords_prods['order_hour_of_day'] = ords_prods['order_hour_of_day'].astype('int8')

In [33]:
ords_prods['add_to_cart_order'] = ords_prods['add_to_cart_order'].astype('int8')

In [34]:
ords_prods['reordered'] = ords_prods['reordered'].astype('int8')

In [35]:
ords_prods['aisle_id'] = ords_prods['aisle_id'].astype('int16')

In [36]:
ords_prods['department_id'] = ords_prods['department_id'].astype('int8')

In [37]:
ords_prods['prices'] = ords_prods['prices'].astype('float16')

In [38]:
ords_prods['days_since_last_order_placed'] = ords_prods['days_since_last_order_placed'].astype('float16')

In [39]:
ords_prods['avg_price'] = ords_prods['avg_price'].astype('float16')

In [40]:
ords_prods['median_order_frequency'] = ords_prods['median_order_frequency'].astype('float16')

In [41]:
#Verifying user_id is now a string 
ords_prods.dtypes

order_id                          int64
user_id                          object
order_number                       int8
orders_day_of_week                 int8
order_hour_of_day                  int8
days_since_last_order_placed    float16
product_id                        int16
add_to_cart_order                  int8
reordered                          int8
product_name                     object
aisle_id                          int16
department_id                      int8
prices                          float16
price_range_loc                  object
busiest_day                      object
busiest_days                     object
Most_orders                      object
max_order                         int64
loyalty_flag                     object
avg_price                       float16
spending_flag                    object
median_order_frequency          float16
frequency_flag                   object
dtype: object

#Merging customer.csv dataframe into orders_products_merge_complete.pkl

In [42]:
ords_prods_cust = ords_prods.merge(cust, on ='user_id')

In [43]:
ords_prods_cust.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   

   days_since_last_order_placed  product_id  add_to_cart_order  reordered  \
0                           NaN         196                  1          0   
1                          15.0         196                  1          1   
2                          21.0         196                  1          1   
3                          29.0         196                  1          1   
4                          28.0         196                  1          1   

  product_name  ...         frequency_flag  First Name  Last Name  gender  \
0         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined  no_of_dependents family_status  income  
0  Alabama  31   2/17/2019                 3       married   40423  
1  Alabama  31   2/17/2019                 3       married   40423  
2  Alabama  31   2/17/2019                 3       married   40423  
3  Alabama  31   2/17/2019                 3       married   40423  
4  Alabama  31   2/17/2019                 3       married   40423  

[5 rows x 32 columns]

In [44]:
ords_prods_cust.shape

(32404859, 32)

Ensuring that 9 new columns were added to the merge. There is a total of 10 columns in the original customer.csv but they are merged via the user_id column ( total 9 columns merged). Merge is proven successful. 

7. Ensure your notebook contains logical titles, section headings, and descriptive code comments.
8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

#Exporting Dataframe 

In [45]:
ords_prods_cust.to_pickle(os.path.join(path,'Data', 'Prepared Data', 'complete_instacart_data.pkl'))